### PR TITLE
Hlsl compiler examples

### PR DESCRIPTION
**Pull request by deprilula28**
Uses HLSL for example 02 to show off the HLSL compiler

### CODE REVIEW

File: 04.Keyframe/main.cpp [hunk 1]
```diff
@@ -144,15 +144,17 @@ int main()
...
+			options.preprocessorOptions.sourceIdentifier = "runtimeID";
+			auto spirv = device->getAssetManager()->getCompilerSet()->compileToSPIRV(source, options);
 			return core::make_smart_refctd_ptr<asset::ICPUSpecializedShader>(std::move(spirv),asset::ICPUSpecializedShader::SInfo{ nullptr,nullptr,"main",stage });
 		};
 		// origFilepath is only relevant when you have filesystem #includes in your shader
 		auto createSpecializedShaderFromSourceWithIncludes = [&](const char* source, asset::ISpecializedShader::E_SHADER_STAGE stage, const char* origFilepath)
 		{
-			auto resolved_includes = device->getAssetManager()->getGLSLCompiler()->resolveIncludeDirectives(source, stage, origFilepath);
-			return createSpecializedShaderFromSource(reinterpret_cast<const char*>(resolved_includes->getContent()->getPointer()), stage);
+			return createSpecializedShaderFromSource(source, stage);
```
Comment (Erfan-Ahmadi) [inline 04.Keyframe/main.cpp:157]: how to replace usages of `resolveIncludeDirectives`:

A. If you remove usage of `resolveIncludeDirectives` you should fill in `SCompilerOptions::includeFinder` in order for the compile function to preprocess and resolve the includes. (get the default include finder from the appropriate compiler)

B. if you want to replace the usage of the function, it's just renamed to `preprocessShader` with a few tweaks in the CShaderCompiler function. (get the appropriate compiler based on the language first and then call the preprocessShader function)

Prefer using **A**

File: 35.GeometryCreator/main.cpp [hunk 2]
```diff
@@ -278,7 +278,14 @@ class GeometryCreatorSampleApp : public ApplicationBase
 		auto createSpecializedShaderFromSource = [=](const char* source, asset::IShader::E_SHADER_STAGE stage) -> core::smart_refctd_ptr<video::IGPUSpecializedShader>
 		{
 			// TODO: Update: should use the compiler set from asset manager
```
Comment (Erfan-Ahmadi) [inline 35.GeometryCreator/main.cpp:280]: remove TODO

File: 35.GeometryCreator/main.cpp [hunk 3]
```diff
@@ -288,7 +295,15 @@ class GeometryCreatorSampleApp : public ApplicationBase
 
 		auto createSpecializedShaderFromSourceWithIncludes = [&](const char* source, asset::IShader::E_SHADER_STAGE stage, const char* origFilepath)
 		{
-			auto resolved_includes = assetManager->getGLSLCompiler()->resolveIncludeDirectives(source, stage, origFilepath);
+			auto computeUnspec = core::make_smart_refctd_ptr<asset::ICPUShader>(source, stage, asset::IShader::E_CONTENT_TYPE::ECT_GLSL);
+
+			auto includeFinder = core::make_smart_refctd_ptr<asset::IShaderCompiler::CIncludeFinder>(core::smart_refctd_ptr(system));
```
Comment (Erfan-Ahmadi) [inline 35.GeometryCreator/main.cpp:300]: Wait what? Just use the default include finder of the compiler

File: 02.ComputeShader/compute.hlsl [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+};
+
+[[vk::push_constant]]
+PushConstants u_pushConstants;
+
+[[vk::binding(0, 0)]] RWTexture2D<float4> outImage;
+[[vk::binding(1, 0)]] Texture2D<float4> inImage;
+
+[numthreads(16, 16, 1)]
+void main(uint3 gl_GlobalInvocationID : SV_DispatchThreadID)
```
Comment (Erfan-Ahmadi) [inline 02.ComputeShader/compute.hlsl:20]: rename it to `invocationID`

File: 35.GeometryCreator/main.cpp [hunk 5]
```diff
@@ -288,7 +293,14 @@ class GeometryCreatorSampleApp : public ApplicationBase
 
 		auto createSpecializedShaderFromSourceWithIncludes = [&](const char* source, asset::IShader::E_SHADER_STAGE stage, const char* origFilepath)
 		{
-			auto resolved_includes = assetManager->getGLSLCompiler()->resolveIncludeDirectives(source, stage, origFilepath);
+			auto computeUnspec = core::make_smart_refctd_ptr<asset::ICPUShader>(source, stage, asset::IShader::E_CONTENT_TYPE::ECT_GLSL);
+			auto shaderCompiler = assetManager->getCompilerSet()->getShaderCompiler(asset::IShader::E_CONTENT_TYPE::ECT_GLSL)->getDefaultIncludeFinder();
+			shaderCompiler->addSearchPath(origFilepath, core::make_smart_refctd_ptr<asset::IShaderCompiler::CFileSystemIncludeLoader>(core::smart_refctd_ptr(system)));
```
Comment (Erfan-Ahmadi) [inline 35.GeometryCreator/main.cpp:298]: I think we talked about this and that you shouldn't do it. no need to addSearchPaths

File: 04.Keyframe/main.cpp [hunk 6]
```diff
@@ -142,23 +142,28 @@ int main()
...
 		};
 		// origFilepath is only relevant when you have filesystem #includes in your shader
+		auto system = nbl::core::make_smart_refctd_ptr<nbl::system::CSystemWin32>();
 		auto createSpecializedShaderFromSourceWithIncludes = [&](const char* source, asset::ISpecializedShader::E_SHADER_STAGE stage, const char* origFilepath)
 		{
-			auto resolved_includes = device->getAssetManager()->getGLSLCompiler()->resolveIncludeDirectives(source, stage, origFilepath);
-			return createSpecializedShaderFromSource(reinterpret_cast<const char*>(resolved_includes->getContent()->getPointer()), stage);
+			asset::IShaderCompiler::SCompilerOptions options;
+			options.preprocessorOptions.includeFinder = new asset::IShaderCompiler::CIncludeFinder(system);
+			options.preprocessorOptions.includeFinder->addSearchPath(origFilepath, options.preprocessorOptions.includeFinder->getDefaultFileSystemLoader());
```
Comment (Erfan-Ahmadi) [inline 04.Keyframe/main.cpp:159]: Again, no need to addSearchPaths, I see why you felt like doing this, but it does nothing, the origFilepath are just names for our shader which go into `sourceIdentifier`

But this example seems way out of date so  I'll revert everything.

File: 35.GeometryCreator/main.cpp (hunk 5 above)
Comment (Erfan-Ahmadi) [reply 35.GeometryCreator/main.cpp:298]: fixed it myself

**Comment by devshgraphicsprogramming** [comment]
@Erfan-Ahmadi can we merge this to `master` or not yet?

**Comment by Erfan-Ahmadi** [comment]
> @Erfan-Ahmadi can we merge this to `master` or not yet?

Only thing remaining:
There is a crash on release mode when trying to delay load dxcompiler.dll.
I have let @AnastaZIuk know.

File: 62.CAD/common.hlsl [hunk 7]
```diff
@@ -44,9 +44,10 @@ struct PSInput
 	float4 position : SV_Position;
     [[vk::location(0)]] float4 color : COLOR; 
     [[vk::location(1)]] nointerpolation float4 start_end : COLOR1; 
-    [[vk::location(2)]] nointerpolation uint3 lineWidth_eccentricity_objType : COLOR2; 
+    [[vk::location(2)]] nointerpolation uint4 lineWidth_eccentricity_objType_writeToAlpha : COLOR2;
```
Comment (devshgraphicsprogramming) [inline 62.CAD/common.hlsl:47]: I don't think you need to give them semantics like `: COLOR` with SPIR-V out?

File: 62.CAD/common.hlsl [hunk 8]
```diff
@@ -44,9 +44,10 @@ struct PSInput
 	float4 position : SV_Position;
     [[vk::location(0)]] float4 color : COLOR; 
     [[vk::location(1)]] nointerpolation float4 start_end : COLOR1; 
-    [[vk::location(2)]] nointerpolation uint3 lineWidth_eccentricity_objType : COLOR2; 
+    [[vk::location(2)]] nointerpolation uint4 lineWidth_eccentricity_objType_writeToAlpha : COLOR2; 
 };
 
 [[vk::binding(0,0)]] ConstantBuffer<Globals> globals : register(b0);
 [[vk::binding(1,0)]] StructuredBuffer<DrawObject> drawObjects : register(t0);
+[[vk::binding(2,0)]] RWTexture2D<uint> pseudoStencil : register(u0);
```
Comment (devshgraphicsprogramming) [inline 62.CAD/common.hlsl:52]: how do you add the `coherent` GLSL qualifier in HLSL?

Comment (deprilula28) [reply 62.CAD/common.hlsl:52]: globallycoherent

**Pull request merged**